src/pykeen/losses.py

```diff
@@ -71,87 +107,175 @@ class BCELoss(PointwiseLoss, nn.BCELoss):
+        return self.reduction_operation(self.margin_activation(neg_scores[:, :, None] - pos_scores[:, None, :] + self.margin))
+
+
+class SetwiseLoss(Loss):
+    """Setwise loss functions compare the scores of several triples."""
+
+    def forward(
+        self,
+        scores: torch.FloatTensor,
+        labels: torch.FloatTensor,
+    ) -> torch.FloatTensor:
+        """Evaluate the loss function.
+
+        :param scores: shape: (batch_size, num_triples_per_batch)
+            The triple scores.
+        :param labels: shape: (batch_size, num_triples_per_batch)
+            The labels for each triple, in [0, 1].
+        """
+        raise NotImplementedError
 
 
 class CrossEntropyLoss(SetwiseLoss):
     """Evaluate cross entropy after softmax output."""
 
-    def __init__(self, reduction: str = 'mean'):
-        super().__init__()
-        self.reduction = reduction
-        self._reduction_method = _REDUCTION_METHODS[reduction]
-
     def forward(
         self,
-        logits: torch.FloatTensor,
+        scores: torch.FloatTensor,
         labels: torch.FloatTensor,
-        **kwargs,
     ) -> torch.FloatTensor:  # noqa: D102
         # cross entropy expects a proper probability distribution -> normalize labels
         p_true = functional.normalize(labels, p=1, dim=-1)
+
         # Use numerically stable variant to compute log(softmax)
-        log_p_pred = logits.log_softmax(dim=-1)
+        log_p_pred = scores.log_softmax(dim=-1)
+
         # compute cross entropy: ce(b) = sum_i p_true(b, i) * log p_pred(b, i)
         sample_wise_cross_entropy = -(p_true * log_p_pred).sum(dim=-1)
-        return self._reduction_method(sample_wise_cross_entropy)
+        return self.reduction_operation(sample_wise_cross_entropy)
 
 
-class NSSALoss(SetwiseLoss):
+class NSSALoss(PairwiseLoss):
     """An implementation of the self-adversarial negative sampling loss function proposed by [sun2019]_."""
 
-    def __init__(self, margin: float, adversarial_temperature: float, reduction: str = 'mean') -> None:
-        super().__init__()
-        self.reduction = reduction
+    # TODO: Actually the loss is pointwise. It is only the weighting, which is setwise on the negative triples.
+
+    def __init__(
+        self,
+        margin: float,
+        adversarial_temperature: float,
+        reduction: str = 'mean',
+    ):
+        """
+        Initialize the loss module.
+
+        :param margin:
+            The margin parameter to use for the base loss. Negative scores should be below -margin, and positive ones
+            above +margin.
+        :param adversarial_temperature:
+            The softmax temperature to use for computing the weights of negative scores. Smaller values lead to more
+            uniform distribution, whereas large values in the limit only consider the largest negative score.
+            # TODO: The usage of temperature here is inverse to the "normal" usage of softmax temperature.
```

Comment:
@cthoyt The common terminology for softmax temperature refers to a scalar by which the pre-softmax scores are _divided_. However, in this loss function, the scores get _multiplied_ by this factor, hence making it effectively an inverse temperature.

So, either, we should rename the parameter to `inverse_adversarial_temperature`, and keep all defaults etc, or we divide by it in the `forward` method. In the latter case, we would need to make sure to adapt the default values accordingly, cf. #29 .


It would make more sense to call this parameter `inverse_temperature`

Reply:
Oh boy... these kinds of renames will make our reproducibility study more tricky, so I'm inclined to keep it how it is (can we just do `(1/adversarial_temperature)` in the `__init__()`?

Reply:
but then it's a misnomer still :/

Reply:
The reproducibility study is linked to a very specific version of the repository, so maybe we can still change it now and adapt the configurations?